### PR TITLE
Block some refactorings/fixes on top-level statements

### PR DESCRIPTION
**Pull request by jcouv**
Mitigates a few issues by disabling some fixes/refactorings in presence of top-level statements:
- ~~https://github.com/dotnet/roslyn/issues/44263 (InlineTemp)~~
- https://github.com/dotnet/roslyn/issues/44271 (AddParameter)
- https://github.com/dotnet/roslyn/issues/44260 (ExtractMethod/LocalFunction)

Relates to https://github.com/dotnet/roslyn/issues/43563 (test plan for top-level statements)

### CODE REVIEW

**Comment by jcouv** [comment]
Tagging @CyrusNajmabadi @sharwell @AlekseyTs 
Fixing those scenarios is not trivial given timeline, so I'm adding temporary mitigations.

**Comment by CyrusNajmabadi** [comment]
I'm fine with blocking tihs.  But what tracks reenabling them?  I would totally expect these to work by ship time :)

File: src/EditorFeatures/CSharpTest/AddParameter/AddParameterTests.cs [hunk 1]
```diff
@@ -2587,5 +2589,57 @@ void TestFunc()
...
+    <Project Language=""C#"" LanguageVersion=""Preview"" AssemblyName=""Assembly1"" CommonReferences=""true"">
+        <Document>
+[|local|](1, 2, 3);
+
+void local(int x, int y)
+{
+}
+        </Document>
+    </Project>
+</Workspace>",
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/AddParameter/AddParameterTests.cs:2608]: surprised you used the `<workspace>` form here.  Was that necessary?

File: src/EditorFeatures/CSharpTest/AddParameter/AddParameterTests.cs [hunk 2]
```diff
@@ -2587,5 +2589,57 @@ void TestFunc()
...
+            new TestParameters(parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp8)));
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/AddParameter/AddParameterTests.cs:2609]: i'm confused by the Project xml specifying the lang version and the test specifying something different.

File: src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs [hunk 3]
```diff
@@ -11182,6 +11182,45 @@ private static bool NewMethod()
...
+}";
+            await TestExtractMethodAsync(code, expected);
+        }
+
+        [WorkItem(44260, "https://github.com/dotnet/roslyn/issues/44260")]
+        [Fact, Trait(Traits.Feature, Traits.Features.ExtractMethod)]
+        public async Task TopLevelStatement_ArgumentInInvocation()
+        {
+            // Note: the cast should be simplified 
+            // https://github.com/dotnet/roslyn/issues/44260
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs:11213]: this issue doesn't seem related.

File: src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs [hunk 4]
```diff
@@ -59,6 +59,12 @@ public override async Task ComputeRefactoringsAsync(CodeRefactoringContext conte
                 return;
             }
 
+            if (variableDeclaration.Parent.IsParentKind(SyntaxKind.GlobalStatement))
+            {
+                // We don't handle InlineTemp in top-level statements yet. https://github.com/dotnet/roslyn/issues/44263
+                return;
+            }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs:66]: is this sufficient?  what if you're in a block in the top level?

File: src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs [hunk 5]
```diff
@@ -181,6 +182,14 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
             {
                 if (method.IsNonImplicitAndFromSource())
                 {
+                    var methodNode = root.FindNode(method.Locations[0].SourceSpan);
+                    if (methodNode.RawKind == syntaxFacts.SyntaxKinds.GlobalStatement)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs:186]: this is pretty ugly.  tbh i'm not even sure why this works.  Why does looking for the span of a method end up with the node for a GlocalStatement?

File: src/Features/Core/Portable/ExtractMethod/AbstractExtractMethodService.cs [hunk 6]
```diff
@@ -36,6 +38,15 @@ internal abstract class AbstractExtractMethodService<TValidator, TExtractor, TRe
                 return new FailedExtractMethodResult(selectionResult.Status);
             }
 
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            if (localFunction && syntaxFacts.ContainsGlobalStatement(root))
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ExtractMethod/AbstractExtractMethodService.cs:43]: this is weird. so if i'm extracting to a local function from within a method in a class in a file that hsa global statements, i'm just blocked?

File: src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs (hunk 3 above)
Comment (jcouv) [reply src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs:11213]: The cast gets correctly simplified for similar situation inside in a method, so the problem may be specific to top-level statement somehow.

**Comment by jcouv** [comment]
> But what tracks reenabling them? 

The linked issues will remain open.

File: src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs (hunk 5 above)
Comment (jcouv) [reply src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs:186]: The reason for this specific code is that later in the fixer we use this logic (`root.FindNode(method.Locations[0]...`, see `AddParameterService.cs` line 104) to find the parameter list to update. But that fails because we find a top-level statement.

The `.Locations[0]` seemed correct (it's the identifier/name of the local function). 
But the `FindNode` logic has tie-breaker rule and flag `getInnermostNodeForTie`.

I suspect the ultimate fix might be using `getInnermostNodeForTie=true`, but didn't have time to explore fully.

File: src/Features/Core/Portable/ExtractMethod/AbstractExtractMethodService.cs (hunk 6 above)
Comment (jcouv) [reply src/Features/Core/Portable/ExtractMethod/AbstractExtractMethodService.cs:43]: At the moment yes. I agree it's weird experience :-/

`MethodExtractor.CodeGenerator.GenerateAsync` (line  96-100, where we handle local function scenario) and `CSharpCodeGenerationService.AddStatement` (line 467) is where the problem seems to be, But the logic there is complex enough that I couldn't extract/replicate a simple condition for when it won't work.

File: src/EditorFeatures/CSharpTest/AddParameter/AddParameterTests.cs (hunk 1 above)
Comment (jcouv) [reply src/EditorFeatures/CSharpTest/AddParameter/AddParameterTests.cs:2608]: I didn't know it's possible to using XML approach starting with `<Project>` directly. Will simplify.

Comment (jcouv) [reply src/EditorFeatures/CSharpTest/AddParameter/AddParameterTests.cs:2608]: Hm, on second thought, when I removed `<Workspace>` in test  `TopLevelStatement_Nested` below, it started outputting XML instead of just source. I'm not sure how this works. Is it okay to leave `<Workspace>` here?

File: src/EditorFeatures/CSharpTest/AddParameter/AddParameterTests.cs (hunk 2 above)
Comment (jcouv) [reply src/EditorFeatures/CSharpTest/AddParameter/AddParameterTests.cs:2609]: Yes, sorry about that. 
I can remove the `.WithLanguageVersion` (tests would still work). 
But since it's the main change I'd make to address feedback and the PR has already gone through CI, I'm tempted to leave as-is until the scenarios get properly fixed. Let me know if that's okay.

**Comment by jcouv** [comment]
It looks like @sharwell has a solution ready for InlineTemp, so I'm reverting that portion of the PR.
